gems/sorbet-runtime/lib/types/private/methods/decl_builder.rb

```diff
@@ -167,6 +198,10 @@ def overridable
         raise BuilderError.new(".overridable cannot be repeated in a single signature")
       end
 
+      if blk
+        T::Private::DeclState.current.active_declaration.blk = blk
+      end
+
       self
     end
 
```

Comment:
You are missing cases in ~all the other DeclBuilder methods that prevent those builders from used outside the block. For example, `returns`, `void`, `params`, `checked`, `on_failure`, etc. should all be banned outside the block.

There should be tests for all of these case as well.

Reply:
It would also be nice to have better static errors for misusing these builders outside the block as well. For example:

#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20strict%0A%0Arequire_relative%20'.%2Flib%2Fsorbet-runtime'%0A%0Aextend%20T%3A%3ASig%0A%0Asig.void%0Adef%20foo%0A%20%20puts%20'in%20foo'%0Aend%0A%0Afoo)

```ruby
# typed: strict

require_relative './lib/sorbet-runtime'

extend T::Sig

sig.void
def foo
  puts 'in foo'
end

foo
```

#### Observed output

```
❯ sorbet qux.rb
qux.rb:7: sig requires a block parameter, but no block was passed https://srb.help/7021
     7 |sig.void
        ^^^

qux.rb:8: This function does not have a sig https://srb.help/7017
     8 |def foo
        ^^^^^^^
  Autocorrect: Use `-a` to autocorrect
    qux.rb:8: Insert sig {returns(NilClass)}

     8 |def foo
        ^

qux.rb:7: Method void does not exist on NilClass https://srb.help/7003
     7 |sig.void
        ^^^^^^^^
  Got NilClass originating from:
    qux.rb:3: Possibly uninitialized (NilClass) in:
     3 |require_relative './lib/sorbet-runtime'
     4 |
     5 |extend T::Sig
     6 |
     7 |sig.void
     8 |def foo
     9 |  puts 'in foo'
    10 |end
    11 |
    12 |foo
  Did you mean:
    https://github.com/sorbet/sorbet/tree/master/rbi/core/kernel.rbi#L1560: Kernel#load
    1560 |  def load(filename, arg0=T.unsafe(nil)); end
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Errors: 3
```

#### Expected behavior

None of the errors reported above are particularly helpful in telling the user what they did wrong (putting the builder method chained on the sig, rather than inside the block).

I think we should go out of our way to make good error messages here, because `sig` syntax is one of the first things that trips up new users of Sorbet.

(Sorbet already does not report errors here, but arguably not reporting errors was ~fine before because it was unlikely for people to attempt to chain methods like this. Now that some of the builders are chained outside and some inside, it's more likely that people will trip up.)

Reply:
It's a bit hard to implement these checks in the runtime, because all of these methods are allowed in type aliases, which would be considered outside of the sig block. Sorbet itself uses this often.

```ruby
MyType = T.type_alias { params().... }
```

Would it be okay to catch these only using static checks?